[WIP][Sema] Don't use 'DotSyntaxCallExpr' in non-typechecked synthesized body

rintaro opened the pull request:
`DotSyntaxCallExpr` is supposed to be only used in typechecked AST.
Use `MemberRefExpr` instead.

(This is a groundwork for removing `SanitizeExpr` from CSGen).

rintaro commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1932,7 +1932,7 @@ void AttributeChecker::visitMainTypeAttr(MainTypeAttr *attr) {
   auto *body = BraceStmt::create(context, SourceLoc(), stmts,
                                 SourceLoc(), /*Implicit*/true);
-  func->setBodyParsed(body);
+  func->setBody(body, AbstractFunctionDecl::BodyKind::TypeChecked);
```
This is still not enough. ASTVisitor says:
```
unspecialized reference with polymorphic type <Self where Self : OtherThing, Self : Runnable> (Self.Type) -> () -> ()
(declref_expr implicit type='<Self where Self : OtherThing, Self : Runnable> (Self.Type) -> () -> ()' location=/Users/rintaro/Repositories/swift-oss/swift/test/attr/ApplicationMain/attr_main_struct_from_two_protocols_one_missing.swift:18:1 range=[/Users/rintaro/Repositories/swift-oss/swift/test/attr/ApplicationMain/attr_main_struct_from_two_protocols_one_missing.swift:18:1 - line:18:1] decl=attr_main_struct_from_two_protocols_one_missing.(file).Runnable extension.main()@/Users/rintaro/Repositories/swift-oss/swift/test/attr/ApplicationMain/attr_main_struct_from_two_protocols_one_missing.swift:12:17 [with (substitution_map generic_signature=<Self where Self : OtherThing, Self : Runnable> (substitution Self -> EntryPoint))] function_ref=unapplied)
```

rintaro replied:
Also, setting `BodyKind::TypeChecked` causes:
```
/Users/rintaro/Repositories/swift-oss/swift/test/attr/ApplicationMain/attr_main_struct_available_future.swift:5:10: error: expected error not produced
@main // expected-error {{'main()' is only available in macOS 10.99 or newer}}
     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

slavapestov replied:
It will take a bit more work to build a fully type checked AST here, for example you have to set types on all the Exprs. Maybe it’s best to avoid doing that for now?

slavapestov replied:
The specific verifier error is because you have a DeclRefExpr without substitutions. If you want to pursue building fully checked AST here you’ll need to provide a substitution map.

rintaro replied:
Removed `setType()`/`setInterfaceType()` part from this function. It should be done in constraint system.

slavapestov commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -647,29 +647,20 @@ deriveBodyEncodable_encode(AbstractFunctionDecl *encodeDecl, void *) {
-    auto *encodeDeclRef = new (C) DeclRefExpr(ConcreteDeclRef(encodeDecl),
-                                              DeclNameLoc(), /*Implicit=*/true);
+    // container.superEncoder
+    auto *superEncoderRef = new (C) UnresolvedDotExpr(containerExpr, SourceLoc(), DeclNameRef(C.Id_superEncoder), DeclNameLoc(), /*Implicit=*/true);
```
These lines should be < 80 columns

rintaro committed:
[Sema] Don't use 'DotSyntaxCallExpr' in non-typechecked synthesized body

`DotSyntaxCallExpr` is supposed to be only used in typechecked AST.
Use `MemberRefExpr` instead.

rintaro committed:
[Experiment] Don't run SanitizeExpr

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please test compiler performance

rintaro commented on lib/Sema/DerivedConformanceComparable.cpp:
```diff
@@ -100,13 +93,11 @@ deriveBodyComparable_enum_noAssociatedValues_lt(AbstractFunctionDecl *ltDecl,
 
   BraceStmt *body = BraceStmt::create(C, SourceLoc(), statements, SourceLoc());
-  return { body, /*isTypeChecked=*/true };
+  return {body, /*isTypeChecked=*/false};
```
Changed `isTypechecked` to `false` because this synthesizer is also used from `deriveBodyComparable_enum_hasAssociatedValues_lt()` below, which is non-typechecked.

rintaro commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -647,31 +647,24 @@ deriveBodyEncodable_encode(AbstractFunctionDecl *encodeDecl, void *) {
-    auto *encodeCall = new (C) DotSyntaxCallExpr(superRef, SourceLoc(),
-                                                 encodeDeclRef);
+    auto *encodeRef = new (C) MemberRefExpr(superRef, SourceLoc(), encodeDecl,
+                                            DeclNameLoc(), /*Implicit=*/true);
```
I feel like this original `DotSyntaxCallExpr(superRef, SourceLoc(),  encodeDeclRef)` is not right.
Shouldn't be `DotSyntaxCallExpr(encodeDeclRef, SourceLoc(), superRef)`?

brentdax replied:
> Shouldn't it be `DotSyntaxCallExpr(encodeDeclRef, SourceLoc(), superRef)`?

Ouch. Yes, it looks like it should be.

rintaro commented:
@swift-ci Please Test Source Compatibility

rintaro commented on lib/Sema/CSGen.cpp:
```diff
@@ -4151,7 +4151,7 @@ namespace {
                                     DeclContext *DC) {
   // Remove implicit conversions from the expression.
-  expr = expr->walk(SanitizeExpr(cs));
+  // expr = expr->walk(SanitizeExpr(cs));
```
Just an experiment. Seeing the effect of removing `SanitizeExpr`

rintaro commented:
@swift-ci Please Test Source Compatibility

rintaro commented:
@swift-ci Please Sourcekit Stress test

rintaro commented:
superseded by #32567

slavapestov commented:
I'm sorry @rintaro, I forgot about this change. Mine is pretty much identical.
